// @progress/kendo-react-layout/appbar/AppBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import o from "prop-types";
import { validatePackage as M, focusFirstFocusableChild as C, useId as g, classNames as y } from "@progress/kendo-react-common";
import { packageMetadata as N } from "../package-metadata.mjs";
const l = e.forwardRef((t, m) => {
  M(N);
  const {
    children: d,
    className: p,
    style: u,
    id: f
  } = t, s = e.useRef(null), c = e.useCallback(
    () => {
      s.current && C(s.current);
    },
    []
  ), k = e.useCallback(
    () => ({
      element: s.current,
      focus: c
    }),
    [c]
  );
  e.useImperativeHandle(m, k);
  const b = g(), a = e.useMemo(
    () => t.themeColor || n.themeColor,
    [t.themeColor]
  ), r = e.useMemo(
    () => t.position || n.position,
    [t.position]
  ), i = e.useMemo(
    () => t.positionMode || n.positionMode,
    [t.positionMode]
  ), h = e.useMemo(
    () => y(
      "k-appbar",
      {
        "k-appbar-top": r === "top",
        "k-appbar-bottom": r === "bottom",
        "k-appbar-static": i === "static",
        "k-appbar-sticky": i === "sticky",
        "k-appbar-fixed": i === "fixed",
        "k-appbar-light": a === "light",
        "k-appbar-dark": a === "dark",
        "k-appbar-inherit": a === "inherit"
      },
      p
    ),
    [r, i, a, p]
  );
  return /* @__PURE__ */ e.createElement(
    "div",
    {
      className: h,
      style: u,
      id: f || b
    },
    d
  );
});
l.propTypes = {
  children: o.any,
  className: o.string,
  style: o.object,
  id: o.string,
  themeColor: o.string,
  position: o.oneOf(["top", "bottom"]),
  positionMode: o.oneOf(["static", "sticky", "fixed"])
};
const n = {
  themeColor: "light",
  position: "top",
  positionMode: "static"
};
l.displayName = "KendoAppBar";
export {
  l as AppBar
};
